core/debug_communication.py

```diff
@@ -0,0 +1,146 @@
+import time
+
+from flask import Flask
+from werkzeug.serving import run_simple
+
+from custom.keras               import get_optimizer
+from models.keras_perceptron    import KerasPerceptron
+from core.utils.dmljob          import DMLJob
+
+
+class FlaskTestingCommunicationManager:
+    """
+    Flask Testing Communication Manager
+
+    This class is for debugging/testing purposes. It implements a communication
+    manager based on Flask so that we can test the UNIX Service functionalities
+    without an active P2P Network.
+
+    """
+
+    def __init__(self, config_manager):
+        """
+        Initializes the instance.
+        """
+        self.config_manager = config_manager
+
+    def configure(self, scheduler):
+        self.listener = FlaskListener(scheduler=scheduler)
+
+    def start(self):
+        """
+        Starts the communication manager (only the listening).
+        """
+        if not self.listener:
+            raise Exception("Communication Manager needs to be configured first.")
+        self.listener.start_listening()
+
+
+class FlaskListener:
+    """
+    Flask Listener
+
+    This class is for debugging/testing purposes. It implements a listener based
+    on Flask so that we can test the UNIX Service functionalities without an
+    active P2P Network.
+
+    """
+
+    def __init__(self, scheduler):
+        """
+        Initializes the listener.
+        """
+        self.scheduler = scheduler
+        self.app = Flask(__name__)
+        self._set_up_endpoints()
+
+    def _set_up_endpoints(self):
+        """
+        Sets up the endpoints of the listener.
+        """
+        self.app.add_url_rule('/', 'heartbeat', self.heartbeat)
+        self.app.add_url_rule('/init_job_async', 'init_job_async', self.init_job_async)
+        self.app.add_url_rule('/init_job_sync', 'init_job_sync', self.init_job_sync)
+        self.app.add_url_rule('/scheduler', 'print_scheduler', self.print_scheduler)
+
+    def start_listening(self):
+        """
+        Starts listening on a port.
+        """
+        run_simple(
+            'localhost',
+            5000,
+            self.app,
+            processes=1,
+            use_reloader=False,
+            use_debugger=False,
+            use_evalex=False
+        )
+
+    def heartbeat(self):
```

Comment:
Yes. 

1. Yes. That line makes the `heartbeat()` function be the function called whenever the `/` endpoint is hit. Not sure if this answers your question.

2. It really doesn't matter what is returned, since the return value of the function is just going to be shown on the browser or returned back as an HTTP Response. Usually the only thing we care about for a heartbeat is whether the HTTP Return Code is successful (in the 200s) or not.